shell/platform/linux/fl_gnome_settings.cc

```diff
@@ -0,0 +1,177 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/linux/fl_gnome_settings.h"
+
+#include <gio/gio.h>
+#include <glib.h>
+
+static constexpr char kDesktopInterfaceSchema[] = "org.gnome.desktop.interface";
+static constexpr char kDesktopTextScalingFactorKey[] = "text-scaling-factor";
+static constexpr char kDesktopClockFormatKey[] = "clock-format";
+static constexpr char kDesktopColorSchemeKey[] = "color-scheme";
+static constexpr char kDesktopGtkThemeKey[] = "gtk-theme";
+
+static constexpr char kClockFormat12Hour[] = "12h";
+static constexpr char kColorSchemePreferDark[] = "prefer-dark";
+static constexpr char kGtkThemeDarkSuffix[] = "-dark";
+
+struct _FlGnomeSettings {
+  GObject parent_instance;
+
+  GSettings* interface_settings;
+};
+
+enum { PROP_0, PROP_INTERFACE_SETTINGS, PROP_LAST };
+
+static void fl_gnome_settings_iface_init(FlSettingsInterface* iface);
+
+G_DEFINE_TYPE_WITH_CODE(FlGnomeSettings,
+                        fl_gnome_settings,
+                        G_TYPE_OBJECT,
+                        G_IMPLEMENT_INTERFACE(fl_settings_get_type(),
+                                              fl_gnome_settings_iface_init))
+
+static FlClockFormat fl_gnome_settings_get_clock_format(FlSettings* settings) {
+  FlGnomeSettings* self = FL_GNOME_SETTINGS(settings);
+
+  FlClockFormat clock_format = FL_CLOCK_FORMAT_24H;
+
+  if (self->interface_settings != nullptr) {
+    g_autofree gchar* value =
+        g_settings_get_string(self->interface_settings, kDesktopClockFormatKey);
+    if (g_strcmp0(value, kClockFormat12Hour) == 0) {
+      clock_format = FL_CLOCK_FORMAT_12H;
+    }
+  }
+  return clock_format;
+}
+
+static bool has_setting(GSettings* settings, const gchar* key) {
+  g_autoptr(GSettingsSchema) schema = nullptr;
+  g_object_get(settings, "settings-schema", &schema, nullptr);
+  return g_settings_schema_has_key(schema, key);
+}
+
+static FlColorScheme fl_gnome_settings_get_color_scheme(FlSettings* settings) {
+  FlGnomeSettings* self = FL_GNOME_SETTINGS(settings);
+
+  FlColorScheme color_scheme = FL_COLOR_SCHEME_LIGHT;
+
+  if (self->interface_settings != nullptr) {
+    if (has_setting(self->interface_settings, kDesktopColorSchemeKey)) {
+      // org.gnome.desktop.interface.color-scheme in GNOME 42 and later
+      g_autofree gchar* value = g_settings_get_string(self->interface_settings,
+                                                      kDesktopColorSchemeKey);
```

Comment:
Is there any particular reason for reading this from the GNOME specific gsettings key instead of using the [org.freedesktop.appearance.color-scheme](https://github.com/flatpak/xdg-desktop-portal/blob/d7a304a00697d7d608821253cd013f3b97ac0fb6/data/org.freedesktop.impl.portal.Settings.xml#L33-L45) settings portal?

For what it's worth, this is GNOME's recommendation [for other toolkits](https://gitlab.gnome.org/GNOME/Initiatives/-/wikis/Dark-Style-Preference#other):
> If it's not possible to use libadwaita or libhandy, use DBus to manually access the settings portal and fetch the preference.
> 
> Do *not* use the gsettings key backing the settings portal implementation other than as a fallback if the portal isn't present: that key only exists on GNOME and is an implementation detail.

The [linked SDL example](https://gitlab.gnome.org/GNOME/Initiatives/-/wikis/uploads/b590f6ee98c68b652d95708f707ce927/hello.c) from that page makes it look relatively straight forward to implement this in a desktop environment agnostic way.

Reply:
Thanks for the suggestion! I'll start looking into this again now that the dependencies have been merged.

Reply:
Works great, here running on KDE. GDBusProxy is unfortunately not mockable/testable so I still have some things left to make this code testable, but otherwise, things are starting to look good. 

![Screenshot from 2022-05-04 10-45-37](https://user-images.githubusercontent.com/140617/166649406-095a1f9d-8abe-4564-a718-8463fa71048a.png)
![Screenshot from 2022-05-04 10-45-07](https://user-images.githubusercontent.com/140617/166649417-fac5d126-baf9-407e-96b1-8f674f6542e7.png)

https://github.com/jpnurmi/engine/commits/settings-portal